File: csharp/Tachyon.Sdk/DocumentsResource.cs
namespace Tachyon.Sdk;

/// <summary><c>/collections/{name}/documents</c> — index, fetch, and delete documents.</summary>
public sealed class DocumentsResource
{
    private readonly HttpTransport _transport;
    private readonly string _collectionName;

    internal DocumentsResource(HttpTransport transport, string collectionName)
    {
        _transport = transport;
        _collectionName = collectionName;
    }

    /// <summary>
    /// <c>POST /collections/{name}/documents</c>. Upserts one or more
    /// documents by id. Individual documents can fail without failing
    /// their neighbours — check <see cref="DocumentsIndexResponse.NumFailed"/>
    /// and <see cref="DocumentsIndexResponse.Results"/>.
    /// </summary>
    public Task<DocumentsIndexResponse> IndexAsync(params Document[] documents) =>
        IndexAsync(documents, CancellationToken.None);

    public Task<DocumentsIndexResponse> IndexAsync(IReadOnlyList<Document> documents, CancellationToken cancellationToken = default) =>
        _transport.RequestAsync<DocumentsIndexResponse>(
            HttpMethod.Post,
            $"/collections/{Uri.EscapeDataString(_collectionName)}/documents",
            body: documents,
            cancellationToken: cancellationToken);

    /// <summary><c>GET /collections/{name}/documents/{id}</c>.</summary>
    public Task<Document> RetrieveAsync(string id, CancellationToken cancellationToken = default) =>
        _transport.RequestAsync<Document>(
            HttpMethod.Get,
            $"/collections/{Uri.EscapeDataString(_collectionName)}/documents/{Uri.EscapeDataString(id)}",
            cancellationToken: cancellationToken);

    /// <summary><c>DELETE /collections/{name}/documents/{id}</c>.</summary>
    public Task DeleteAsync(string id, CancellationToken cancellationToken = default) =>
        _transport.RequestAsync(
            HttpMethod.Delete,
            $"/collections/{Uri.EscapeDataString(_collectionName)}/documents/{Uri.EscapeDataString(id)}",
            cancellationToken: cancellationToken);
}

File: java/src/main/java/io/github/adikeshri/tachyon/DocumentsResource.java
package io.github.adikeshri.tachyon;

import java.util.List;

/** {@code /collections/{name}/documents} — index, fetch, and delete documents. */
public final class DocumentsResource {
    private final HttpTransport transport;
    private final String collectionName;

    DocumentsResource(HttpTransport transport, String collectionName) {
        this.transport = transport;
        this.collectionName = collectionName;
    }

    /**
     * {@code POST /collections/{name}/documents}. Upserts one or more
     * documents by id. Individual documents can fail without failing their
     * neighbours — check {@link DocumentsIndexResponse#numFailed} and
     * {@link DocumentsIndexResponse#results}.
     */
    public DocumentsIndexResponse index(Document... documents) {
        return index(List.of(documents));
    }

    public DocumentsIndexResponse index(List<Document> documents) {
        String path = "/collections/" + UrlEncoding.pathSegment(collectionName) + "/documents";
        return transport.request("POST", path, null, documents, DocumentsIndexResponse.class);
    }

    /** {@code GET /collections/{name}/documents/{id}}. */
    public Document retrieve(String id) {
        String path = "/collections/" + UrlEncoding.pathSegment(collectionName) + "/documents/" + UrlEncoding.pathSegment(id);
        return transport.request("GET", path, null, null, Document.class);
    }

    /** {@code DELETE /collections/{name}/documents/{id}}. */
    public void delete(String id) {
        String path = "/collections/" + UrlEncoding.pathSegment(collectionName) + "/documents/" + UrlEncoding.pathSegment(id);
        transport.requestVoid("DELETE", path);
    }
}
